toolchain/lower/file_context.cpp

```diff
@@ -371,13 +410,12 @@ auto FileContext::BuildFunctionDefinition(SemIR::FunctionId function_id)
     function_lowering.LowerBlock(block_id);
   }
```

Comment:
This unconditionally creates an entry block separate from the function body, which seems a little wasteful. Perhaps instead, we could allow the decl block to fall through into the entry block:
```c++
  for (auto block_id : llvm::concat<const SemIR::InstBlockId>(
           singleton_decl_block, body_block_ids)) {
    CARBON_VLOG("Lowering {0}\n", block_id);
    // Reuse the decl block as the first block of the body if possible. We assume that
    // the first body block is never a branch target.
    if (block_id != body_block_ids.front() || !function_lowering.builder().GetInsertBlock()) {
      auto* llvm_block = function_lowering.GetBlock(block_id);
      // Keep the LLVM blocks in lexical order.
      llvm_block->moveBefore(llvm_function->end());
      function_lowering.builder().SetInsertPoint(llvm_block);
    }
    function_lowering.LowerBlock(block_id);
  }
```

Reply:
Hmm. The code I deleted just below here suggests that we can't assume the first body block is never a branch target, but how about this?

Reply:
Thanks, that seems like it strikes a good balance.